core/src/commonMain/kotlin/Observation.kt

```diff
@@ -0,0 +1,101 @@
+package com.juul.kable
+
+import com.juul.kable.State.Connecting.Observes
+import com.juul.kable.logs.Logger
+import com.juul.kable.logs.Logging
+import kotlinx.atomicfu.atomic
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal class Observation(
+    private val state: StateFlow<State>,
+    private val handler: Handler,
+    private val characteristic: Characteristic,
+    logging: Logging,
+    identifier: String,
+) {
+
+    interface Handler {
+        suspend fun startObservation(characteristic: Characteristic)
+        suspend fun stopObservation(characteristic: Characteristic)
+    }
+
+    private val logger = Logger(logging, tag = "Kable/Observation", identifier)
+
+    private val mutex = Mutex()
+    private val subscribers = mutableListOf<OnSubscriptionAction>()
+    private val isObservationEnabled = atomic(false)
+
+    private val isConnected: Boolean
+        get() = state.value.isAtLeast<Observes>()
+
+    private val hasSubscribers: Boolean
+        get() = subscribers.isNotEmpty()
+
+    suspend fun onSubscription(action: OnSubscriptionAction) = mutex.withLock {
+        subscribers += action
+        enableObservationIfNeeded()
+        if (isObservationEnabled.value) {
+            suppressConnectionExceptions {
+                action()
+            }
+        }
+    }
+
+    suspend fun onCompletion(action: OnSubscriptionAction) = mutex.withLock {
+        subscribers -= action
+        disableObservationIfNeeded()
+    }
+
+    suspend fun onConnected() = mutex.withLock {
+        // `force = true` is used in case a disconnected state (to reset the observation state) was missed.
+        enableObservationIfNeeded(force = true)
```

Comment:
Interesting, am curious what case requires this. Is there a specific unit test that replicates the asymmetric observations enable/disable that fails if `force == false`?

Reply:
I think the function name is a bit misleading, and isn't helpful to understand why.

I'll try to explain the reasoning and maybe you'll have an idea of better naming or better approach:

Essentially, we need to start observations (via platform specific calls) when all of the following conditions are met:
- we are connected
- subscribers has gone from `0` to `1`

Typically this is triggered in `enableObservationIfNeeded` via:

https://github.com/JuulLabs/kable/blob/b8fd68a8266f384e859cd8bbbdb14f65619b4ee7/core/src/commonMain/kotlin/Observation.kt#L67

| Condition | Logical check |
|-|-|
| we are connected | `isConnected` |
| subscribers has gone from `0` to `1` | `!isObservationEnabled.value && hasSubscribers` |

But...and this is where `force` comes in: If we lose connection, then on reconnect, we will **always** want to start observation if we `hasSubscribers` (regardless if we've already started previously, i.e. even if `isObservationEnabled.value` is true). This is because when the connection drops, all platforms clear the observations for us. There isn't a way to know that has happened other than monitoring for disconnected events, hence the `onConnectionLost` function — but that isn't 100% reliable as it is coming from a conflated channel. So, the approach is to say that if `onConnected` is called then we are certain this is a _new_ connection, so it is safe to always start up the observations again (i.e. `force`).

It might be slightly clearer to re-write the `if` statement as?:

```
if ((force || !isObservationEnabled.value) && isConnected && hasSubscribers) {
```

Showing that `force` is overriding the `!isObservationEnabled.value` check.

Reply:
In hindsight, maybe checking that `subscribers.size == 1` would be better than `!isObservationEnabled.value && hasSubscribers`?

For some reason, the `isObservationEnabled` boolean felt safer? But, looking at it now, it just seems to be adding unnecessary complexity.

Reply:
> Interesting, am curious what case requires this. Is there a specific unit test that replicates the asymmetric observations enable/disable that fails if `force == false`?

To answer your initial comment: any test that starts observations on connect (when we already have subscribers) is essentially utilizing `force = true`, and would fail if `force = false`.

Reply:
Significantly simplified the code. Removed the need for `force`, and figured out that having special handling on the disconnected event was unnecessary.